
project vendor/mk/
diff --git a/vendor/mk/config/common.mk b/vendor/mk/config/common.mk
index 2b5aa6b..5f46674 100644
--- a/vendor/mk/config/common.mk
+++ b/vendor/mk/config/common.mk
@@ -3,7 +3,7 @@ PRODUCT_BRAND ?= mokee
 SUPERUSER_EMBEDDED := true
 SUPERUSER_PACKAGE_PREFIX := com.android.settings.cyanogenmod.superuser
 
-MOKEEHELPER_EMBEDDED := true
+MOKEEHELPER_EMBEDDED := false
 MOKEEHELPER_PACKAGE_PREFIX := com.android.settings.mokee.mokeehelper
 
 ifneq ($(TARGET_SCREEN_WIDTH) $(TARGET_SCREEN_HEIGHT),$(space))
@@ -110,30 +110,13 @@ PRODUCT_COPY_FILES += \
 # MoKee prebuilt
 PRODUCT_COPY_FILES += \
     vendor/mk/prebuilt/ota/verifier:system/bin/verifier \
-    vendor/mk/prebuilt/common/etc/init.d/88preinstall:system/etc/init.d/88preinstall \
-    vendor/mk/prebuilt/common/app/iFlyIME.apk:system/app/iFlyIME.apk
+    vendor/mk/prebuilt/common/etc/init.d/88preinstall:system/etc/init.d/88preinstall
 
 # Use all prebuilt lib files
 PRODUCT_COPY_FILES += $(shell test -d vendor/mk/prebuilt/common/lib && \
     find vendor/mk/prebuilt/common/lib -name '*.so' \
     -printf '%p:system/lib/%f ')
 
-# Use all developers-party apk
-PRODUCT_COPY_FILES += $(shell test -d vendor/mk/prebuilt/$(DEVELOPER_MAINTAINER)/app && \
-    find vendor/mk/prebuilt/$(DEVELOPER_MAINTAINER)/app -name '*.apk' \
-    -printf '%p:system/third-app/%f ')
-
-# Use all third-party apk
-PRODUCT_COPY_FILES += $(shell test -d vendor/mk/prebuilt/third/app && \
-    find vendor/mk/prebuilt/third/app -name '*.apk' \
-    -printf '%p:system/third-app/%f ')
-
-# Google IME
-ifneq ($(TARGET_EXCLUDE_GOOGLE_IME),true)
-PRODUCT_COPY_FILES += \
-    vendor/mk/prebuilt/common/app/GoogleIME.apk:system/app/GoogleIME.apk
-endif
-
 # Bring in camera effects
 PRODUCT_COPY_FILES +=  \
     vendor/mk/prebuilt/common/media/LMprec_508.emd:system/media/LMprec_508.emd \
@@ -188,8 +171,8 @@ PRODUCT_PACKAGES += \
     audio_effects.conf \
     Apollo \
     CMFileManager \
+    LatinIME \
     LockClock \
-    MoKeeHelper \
     MoKeeHome \
     MoKeeScreenRecorder \
     MoKeeSetupWizard
@@ -207,7 +190,6 @@ PRODUCT_PACKAGES += \
     mke2fs \
     tune2fs \
     bash \
-    nano \
     htop \
     powertop \
     lsof \
diff --git a/vendor/mk/config/common_full.mk b/vendor/mk/config/common_full.mk
index 8cd6d33..96e0f84 100644
--- a/vendor/mk/config/common_full.mk
+++ b/vendor/mk/config/common_full.mk
@@ -22,6 +22,5 @@ PRODUCT_PACKAGES += \
 
 # Extra tools in MK
 PRODUCT_PACKAGES += \
-    vim \
     zip \
-    unrar
\ No newline at end of file
+    unrar
