
project frameworks/base/
diff --git a/frameworks/base/packages/SystemUI/res/menu/recent_popup_menu.xml b/frameworks/base/packages/SystemUI/res/menu/recent_popup_menu.xml
index 3b36d8a..9b605f9 100644
--- a/frameworks/base/packages/SystemUI/res/menu/recent_popup_menu.xml
+++ b/frameworks/base/packages/SystemUI/res/menu/recent_popup_menu.xml
@@ -18,7 +18,6 @@
 */
 -->
 <menu xmlns:android="http://schemas.android.com/apk/res/android">
-    <item android:id="@+id/recent_add_split_view" android:title="@string/status_bar_recent_add_to_split_view" />
     <item android:id="@+id/recent_remove_item" android:title="@string/status_bar_recent_remove_item_title" />
     <item android:id="@+id/recent_inspect_item" android:title="@string/status_bar_recent_inspect_item_title" />
     <item android:id="@+id/recent_force_stop" android:title="@string/advanced_dev_option_force_stop" />
diff --git a/frameworks/base/packages/SystemUI/src/com/android/systemui/recent/RecentsPanelView.java b/frameworks/base/packages/SystemUI/src/com/android/systemui/recent/RecentsPanelView.java
index 7a6a340..a0eec13 100644
--- a/frameworks/base/packages/SystemUI/src/com/android/systemui/recent/RecentsPanelView.java
+++ b/frameworks/base/packages/SystemUI/src/com/android/systemui/recent/RecentsPanelView.java
@@ -1307,10 +1307,10 @@ public class RecentsPanelView extends FrameLayout implements OnItemClickListener
                     } else {
                         throw new IllegalStateException("Oops, no tag on view " + selectedView);
                     }
-                } else if (item.getItemId() == R.id.recent_add_split_view) {
+            /*    } else if (item.getItemId() == R.id.recent_add_split_view) {
                     // Either start a new activity in split view, or move the current task
                     // to front, but resized
-                    openInSplitView(viewHolder, -1);
+                    openInSplitView(viewHolder, -1); */
                 } else {
                     return false;
                 }
diff --git a/frameworks/base/services/java/com/android/server/wm/WindowManagerService.java b/frameworks/base/services/java/com/android/server/wm/WindowManagerService.java
index 0e7a46c..02dbcc4 100644
--- a/frameworks/base/services/java/com/android/server/wm/WindowManagerService.java
+++ b/frameworks/base/services/java/com/android/server/wm/WindowManagerService.java
@@ -11231,7 +11231,7 @@ public class WindowManagerService extends IWindowManager.Stub
     public Rect getSplitViewRect(int taskId, boolean resetLocation) {
         Log.e("XPLOD", "Debug Activity " + taskId + " asked for split view rect");
         mSplitViewTasks[mNextSplitViewLocation] = taskId;
-        mIsTaskSplitted.put(taskId, true);
+        mIsTaskSplitted.put(taskId, false);
 
         // TODO(multidisplay): For now, apply Configuration to main screen only.
         final DisplayContent displayContent = getDefaultDisplayContentLocked();
