
project build/
diff --git a/build/envsetup.sh b/build/envsetup.sh
index a1cc695..10e93fc 100644
--- a/build/envsetup.sh
+++ b/build/envsetup.sh
@@ -624,7 +624,7 @@ function lunch()
         export CCACHE_DIR=$TMP_CCACHE_DIR/$product
         export CCACHE_BASEDIR=$ANDROID_BUILD_TOP
         if [ -z "$CCACHE_SIZE" ]; then
-            CCACHE_SIZE=16G
+            CCACHE_SIZE=50G
         fi
         if [ "$(uname)" = "Darwin" ] ; then
             prebuilts/misc/darwin-x86/ccache/ccache -M $CCACHE_SIZE
